### ISSUE TITLE
Using DOT_PATH with a symlink for dot does not always work

### ISSUE THREAD

**Issue by kzrkk** [doxygen/doxygen#9007]
**The bug**
Ubuntu 20.04 LTS, set the path to dot in the configuration:

DOT_PATH = /usr/bin

Now call doxygen to create the documentation, dot execution fails, with error "/usr/sbin/dot: not found"

If DOT_PATH is not set and dot is only found via PATH, everything works as expected.

**Expected behaviour**
The error in case of an explicit DOT_PATH should not occur.

**Problem details**
The problem is a result of the fact, that in the mentioned Ubuntu version, /usr/bin/dot is a symlink to /usr/sbin/libgvc6-config-update. Now, within src/configimpl.l DOT_PATH is processed such that `FileInfo::dirPath` is used to get the path to the executable. This funtion in turn calls in the end `ghc::filesystem::canonical()` which in turn follows the symlinks and returns the link destination.

Now we have *DOT_PATH=/usr/bin* vs. the resolved path */usr/sbin*. In the end *dot* as executable name is appended again to the resolved path, but in */usr/sbin* there is no *dot* so the execution fails.

What is the reason for resolving the symlinks here? This way it is impossible to have an actual executable for *dot* that is named differently and that does not reside in the same directory as the *dot* symlink. This looks like an unnecessary restriction to me. Especially because by not setting DOT_PATH at all and letting *dot* to be taken from PATH it works. If someone stumbles across this problem he is unable to understand the root cause without debugging the doxygen code. There is no message whatsoever that the provided path /usr/bin is changed for some reason, you only see that eventually /usr/sbin is used.

The same problem might also exist for other path settings, I have not checked that.

**Version**
Ubuntu 20.04.3 LTS, Doxygen version 1.9.4 built from git:
1.9.4 (dc2950946033e9dab125e389af6eafc3d1f73eac)

**Comment by albert-github** [comment]
The `DOT_PATH` signals the path where the `dot` executable can be found, there has not been the intention that another "dot" can be used, even worse when an executable is given and it is different from `dot` it might be able to find it but later on the other executable name is stripped and an explicit `dot` is appended.
The reason for checking whether or not the `dot` executable exists is that a proper warning is given like:
```
warning: the dot tool could not be found at ...\mydot.exe
```

When you run `doxygen -d extcmd` what do you see when you have `DOT_PATH ` set to `/usr/bin`?
Do you mean that `libgvc6-config-update` is a sub-directory of `/usr/sbin`? and that the resulting executable name should be `/usr/sbin/libgvc6-config-update/dot` ?

**Comment by doxygen** [comment]
I checked on my Ubuntu system and see the same thing:

```
ls -al /usr/bin/neato /usr/bin/dot /usr/bin/circo
lrwxrwxrwx 1 root root  3 Mar  2  2020 /usr/bin/circo -> dot
lrwxrwxrwx 1 root root 29 Mar  2  2020 /usr/bin/dot -> ../sbin/libgvc6-config-update
lrwxrwxrwx 1 root root  3 Mar  2  2020 /usr/bin/neato -> dot
```

Apparently the executable `/usr/sbin/libgvc6-config-update` is used as a generic executable that is aliased by a symbolic links. This executable will probably look at how how it is started (`argv[0]`) and act accordingly.

By resolving the symlink this no longer works.

Two possible workarounds:
1. make sure dot can be found in the path  and keep `DOT_PATH` empty.
2. run doxygen from a wrapper script like this:

```
cp /usr/sbin/libgvc6-config-update dot
cat Doxyfile; echo DOT_PATH='$(PWD)/dot' | doxygen
rm dot
```

**Comment by kzrkk** [comment]
The proposed workaround 1. is what I use now, but this forces me to use different configurations depending on the platform (more specifically the type of the doxygen executable).

Anyway, with doxygen 1.8.17 (from Ubuntu 20.04) the problem does not exist, there `DOT_PATH = /usr/bin` works and leads to calling `/usr/bin/dot`.

So this is one more reason to consided this a bug. If the current behaviour is really intended, at least the description of `DOT_PATH` should be extended to mention the restrictions (no symlinks or at least the symlink target in the same directory as the `dot` symlink.

**Comment by albert-github** [comment]
I think the relevant part of the code (in configimpl.l is:
```
  // check dot path
  QCString dotPath = Config_getString(DOT_PATH);
  if (!dotPath.isEmpty())
  {
    FileInfo fi(dotPath.str());
    if (fi.exists() && fi.isFile()) // user specified path + exec
    {
      dotPath=fi.dirPath(TRUE)+"/";
    }
    else
    {
      QCString dotExe = dotPath+"/dot"+Portable::commandExtension();
      FileInfo dp(dotExe.str());
      if (!dp.exists() || !dp.isFile())
      {
        warn_uncond("the dot tool could not be found at %s\n",qPrint(dotPath));
        dotPath="";
      }
      else
      {
        dotPath=dp.dirPath(TRUE)+"/";  // <==
      }
    }
#if defined(_WIN32) // convert slashes
    uint i=0,l=dotPath.length();
    for (i=0;i<l;i++) if (dotPath.at(i)=='/') dotPath.at(i)='\\';
#endif
    Config_updateString(DOT_PATH,dotPath);
  }
```

when you run your code (without any workaround) I assume you land in the last `dotPath=dp.dirPath(TRUE)+"/";` (i.e. the one I marked  `<==`).

Is this correct?
(I cannot test it as I don't have an Ubuntu or similar system at hand)

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9014 


@kzrkk Note: I could not test it on an Ubuntu system so it is highly appreciated if you could test it.

**Comment by kzrkk** [comment]
@albert-github : Yes, that's where I land.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by kzrkk** [comment]
Unfortunately the fix introduces a new bug if **PlantUML** is used.
If `DOT_PATH` is not set, plantuml is called with the option `-graphvizdot "dot"` which in turn causes plantuml to fail: 
```
Dot Executable: dot
File does not exist
Cannot find Graphviz. ...
```
If `-graphvizdot` is used, it needs a full path (absolute or relative). At least my version of plantuml (1.2021.14) requires this. I am not sure whether this is a bug or a feature of plantuml.

But in the end use of dot alone now works, but use of dot from plantuml not anymore.

**Comment by albert-github** [comment]
Indeed an unexpected regression. Actually my personal opinion is that it is a bug in plantuml, but I think that doxygen should cope with it as well.

**Comment by albert-github** [comment]
For the plantuml problem (https://github.com/doxygen/doxygen/issues/9007#issuecomment-1007651759): I've just pushed a proposed patch, pull request #9018

**Comment by albert-github** [comment]
(Re-implemented) Code for the plantuml problem has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by aminya** [comment]
Is there a download link that contains this fix? Without dot, my documentation looks incorrect.

**Comment by albert-github** [comment]
In the GitHub Actions artifacts of the last master build (see https://github.com/doxygen/doxygen/actions and for the current last master build: https://github.com/doxygen/doxygen/actions/runs/1843009035) there are a number of executables. In case your platform is not present you will have to build from the sources.

Regarding Downgrading doxygen, I don't think this will help.

**Comment by aminya** [comment]
I need a persistent link so I can add it to setup-cpp.
https://github.com/aminya/setup-cpp

Maybe it is time to tag a new release! This seems like an important bug to me.

I remember that the version shipping with ubuntu:devel (1.9.1) was working fine. 
https://packages.ubuntu.com/jammy/doxygen

**Comment by albert-github** [comment]
I don't know if the mentioned bug is important, it is for you for sure but for others all other bugs are important as well.

You could try an older doxygen version but you will have there a number of bugs as well. The problem might also be dependent on the used plantuml version.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
